File: trunk/src/cgi_python/msg_cgi/cgi_msg_get_msg_list_by_session.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

import sys
import time

# 包含上级目录，即../common
sys.path.append('..')

# cgi, proto_io
from common.cgi.cgi_ret_code_def import CGI_RET_CODE
import common.cgi.cgi_my as _cgi
import common.proto_io.proto_io_tcp_client as _tcp_client

# proto
import common.proto.msg_pb2 as _msg
import common.proto.msg_sixin_pb2 as _msg_sixin
import common.proto.comm_struct_pb2 as _comm_struct
import common.proto.comm_enum_pb2 as _comm_enum

# util
from common.util.common_util import *

# msgServerIP = '127.0.0.1'

# 一次拉取消息数量限制
getMsgAmountLimit = 20
# 时间间隔 8*60*60 = 8 hours
interval = 28800

def FetchUserInfo(listOpenidMd5):
	# 1.Build requestMsg protobuf
	stUserInfoRequestMsg = _msg.Msg()

	stUserInfoRequestMsgHead = stUserInfoRequestMsg.head
	stUserInfoRequestMsgHead.cmd = _msg.QUERY_USER_DETAIL_INFO_REQ
	stUserInfoRequestMsgHead.seq = int(time.time())

	stQueryUserDetailInfoReq = stUserInfoRequestMsg.query_user_detail_info_req
	for openidMd5 in listOpenidMd5:
		stQueryUserDetailInfoReq.openid_md5_list.append(openidMd5)

	# 2.Send to server, and recv responseMsg protobuf
	ioclient = _tcp_client.ProtoIOTcpClient(userServerIP, userServerPort)
	dictMsg = {'requestMsg':stUserInfoRequestMsg, 'responseMsg':None, 'strErrMsg':None}

	ret = ioclient.io(dictMsg)

	if ret != 0:	
		logger.error('FetchFeedsInfo IO failed, errmsg = %s', dictMsg['strErrMsg'])
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return -1

	stUserInfoResponseMsg = dictMsg['responseMsg']
	logger.debug('responseMsg: \n %s', stUserInfoResponseMsg)

	if stUserInfoResponseMsg.head.cmd != _msg.QUERY_USER_DETAIL_INFO_RES:
		logger.error('FetchUserInfo response.head.cmd=%d, unkown, fuck!!!', stUserInfoResponseMsg.head.cmd)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return -1

	if stUserInfoResponseMsg.head.result != _msg.E_OK:
		logger.error('FetchUserInfo response.head.result=%d, not E_OK!', stUserInfoResponseMsg.head.result)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return -1

	# 3.Parse	
	res = stUserInfoResponseMsg.query_user_detail_info_res
	for i in range(len(res.user_detail_info_list)):
		if listOpenidMd5[0] == res.user_detail_info_list[i].openid_md5:
			stHostUserInfo = UserInfoPB2Dict(res.user_detail_info_list[i])
			mycgi.addBody('userinfo_from', stHostUserInfo)
			break
		elif i == len(res.user_detail_info_list) - 1:
			return -1

	for i in range(len(res.user_detail_info_list)):
		if listOpenidMd5[1] == res.user_detail_info_list[i].openid_md5:
			stGuestUserInfo = UserInfoPB2Dict(res.user_detail_info_list[i])
			mycgi.addBody('userinfo_to', stGuestUserInfo)
			break
		elif i == len(res.user_detail_info_list) - 1:
			return -1	

	return 0

def InnerProcess():
	# 1.Get HTTP params	
	msg_id = int(mycgi.getEnvValue('msg_id', 0))
	amount = int(mycgi.getEnvValue('amount', 0))
	openid_md5_from = int(mycgi.getEnvValue('openid_md5_from', 0))
	openid_md5_to = int(mycgi.getEnvValue('openid_md5_to', 0))
	
	# 2.Params check
	if 0 == openid_md5_from or 0 == openid_md5_to:
		logger.error('INVALID OPENID MD5! openid_md5_from = %d, openid_md5_to = %d', openid_md5_from, openid_md5_to)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_INVALID_PARAM)		
		return True
	if amount > getMsgAmountLimit:
		logger.error('amount: %d > getMsgAmountLimit: %d', amount, getMsgAmountLimit)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_INVALID_PARAM)		
		return True	

	# 3.Build requestMsg protobuf
	stRequestMsg = _msg.Msg()

	stRequestMsgHeader = stRequestMsg.head
	stRequestMsgHeader.cmd = _msg.QUERY_MSG_REQ
	stRequestMsgHeader.seq = int(time.time())

	stRequestMsgBody = stRequestMsg.msg_query_req
	stRequestMsgBody.msg_id = msg_id
	stRequestMsgBody.amount = amount
	stRequestMsgBody.openid_md5_from = openid_md5_from
	stRequestMsgBody.openid_md5_to = openid_md5_to	

	logger.debug('requestMsg: \n %s', stRequestMsg)

	# 4.Send to server, and recv responseMsg protobuf	
	ioclient = _tcp_client.ProtoIOTcpClient(msgServerIP, msgServerPort)

	dictMsg = {'requestMsg':stRequestMsg, 'responseMsg':None, 'strErrMsg':None}

	ret = ioclient.io(dictMsg)

	if ret != 0:	
		logger.error('ProtoIOTcpClient IO failed, errmsg = %s', dictMsg['strErrMsg'])
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	stResponseMsg = dictMsg['responseMsg']
	logger.debug('responseMsg: \n %s', stResponseMsg)

	if stResponseMsg.head.cmd != _msg.QUERY_MSG_RES:
		logger.error('response.head.cmd=%d, unkown, fuck!!!', stResponseMsg.head.cmd)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	if stResponseMsg.head.result != _msg.E_OK:
		logger.error('response.head.result=%d, not E_OK!', stResponseMsg.head.result)
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	# 5.protobuf -> json, and do reply
	strMsgQueryRes = stResponseMsg.msg_query_res
	msgContentList = []
	preTime = 0
	for i in range(len(strMsgQueryRes.msg_content)):
		dictItem = {}
		msgContent = strMsgQueryRes.msg_content[i]
		dictItem['id'] = str(msgContent.id)
		dictItem['content'] = msgContent.content
		dictItem['openid_md5_from'] = str(msgContent.openid_md5_from)
		dictItem['openid_md5_to'] = str(msgContent.openid_md5_to)

		if i == 0:
			dictItem['create_ts'] = str(msgContent.create_ts)
		else:
			if msgContent.create_ts - preTime >= interval:
				dictItem['create_ts'] = str(msgContent.create_ts)
			else:
				dictItem['create_ts'] = '0'

		preTime = msgContent.create_ts

		msgContentList.append(dictItem)

	# 6.Get UserInfo
	listOpenidMd5 = []
	listOpenidMd5.append(openid_md5_from)
	listOpenidMd5.append(openid_md5_to)
	if FetchUserInfo(listOpenidMd5) < 0:
		logger.error('FetchUserInfo failed!!!')
		mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_SERVER_BUSY)		
		return True

	# 7.Reply
	mycgi.addBody('msg_content_list', msgContentList)	
	mycgi.DoReply(CGI_RET_CODE.CGI_RET_CODE_OK)
	return True



if __name__ == '__main__':
	# __file__ 获取模块所在路径
	mycgi = _cgi.myCgi(__file__, 'logging.conf')

	# logger	
	logger = mycgi.getLogger()

	InnerProcess()	

File: trunk/src/cgi/msg_cgi/cgi_msg_get_msg_list_by_session.cpp
#include "cgi_msg_server.h"
#include "proto_io_tcp_client.h"
#include <time.h>
#include "msg.pb.h"
#include "user_util.h"

class CgiMsgGetListBySession: public CgiMsgServer
{
public:
	CgiMsgGetListBySession():
		CgiMsgServer(0, "config.ini", "logger.properties")
	{

	}

	bool InnerProcess()
	{
		//1.get HTTP params
		uint64_t msg_id = strtoul(((string)GetInput().GetValue("msg_id")).c_str(), NULL, 10);
		uint32_t amount = (uint32_t)GetInput().GetValue("amount"); 
		uint64_t openid_md5_from = strtoul(((string)GetInput().GetValue("openid_md5_from")).c_str(), NULL, 10);
		uint64_t openid_md5_to = strtoul(((string)GetInput().GetValue("openid_md5_to")).c_str(), NULL, 10);

		if(0 == openid_md5_from || 0 == openid_md5_to)
		{
			DoReply(CGI_RET_CODE_INVALID_PARAM);
			LOG4CPLUS_ERROR(logger, "INVALID OPENID MD5");
			return true;
		}

		if(amount > getMsgAmountLimit)
		{
			DoReply(CGI_RET_CODE_INVALID_PARAM);
			LOG4CPLUS_ERROR(logger, "amount="<<amount<<" > getMsgAmountLimit:"<<getMsgAmountLimit);
			return true;
		}
		
				
		//2.Build requestMsg protobuf
		std::string strErrMsg = "";
		::hoosho::msg::Msg stRequestMsg;
		::hoosho::msg::Msg stResponseMsg;
		
		::hoosho::msg::MsgHead* stRequestMsgHead = stRequestMsg.mutable_head();
		stRequestMsgHead->set_cmd(::hoosho::msg::QUERY_MSG_REQ);		
		stRequestMsgHead->set_seq(time(NULL));

		::hoosho::sixin::QuerySixinReq* stMsgQueryReq = stRequestMsg.mutable_msg_query_req();
		stMsgQueryReq->set_msg_id(msg_id);
		stMsgQueryReq->set_amount(amount);
		stMsgQueryReq->set_openid_md5_from(openid_md5_from);
		stMsgQueryReq->set_openid_md5_to(openid_md5_to);

		//3.send to server, and recv responseMsg protobuf
		common::protoio::ProtoIOTcpClient ioclient(msgServerIP, msgServerPort);

		int iRet;
		iRet = ioclient.io(stRequestMsg, stResponseMsg, strErrMsg);

		if(iRet != 0)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "ProtoIOTcpClient IO failed, errmsg = " << strErrMsg);
			return true;
		}

		LOG4CPLUS_DEBUG(logger, "response Msg: \n" << stResponseMsg.Utf8DebugString());	
		//parse response
		const ::hoosho::msg::MsgHead& stResponseMsgHead = stResponseMsg.head();
		if(stResponseMsgHead.cmd() != ::hoosho::msg::QUERY_MSG_RES)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.cmd="<<stResponseMsgHead.cmd()<<", unknown, fuck!!!");
			return true;	
		}

		if(stResponseMsgHead.result() != ::hoosho::msg::E_OK)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.result="<<stResponseMsgHead.result());
			return true;	
		}

		//4.Build strResponse Json from responseMsg protobuf		
		const ::hoosho::sixin::QuerySixinRes& stMsgQueryRes = stResponseMsg.msg_query_res();
		lce::cgi::CAnyValue msg_content_list;		
		uint64_t preTime = 0;
		int count = stMsgQueryRes.msg_content_size()-1;
		for(int i=0; i<=count; i++)
		{
			lce::cgi::CAnyValue item;
			const ::hoosho::commstruct::MsgContent& msgContent = stMsgQueryRes.msg_content(i);
			item["id"] = int_2_str(msgContent.id());			
			item["content"] = msgContent.content();
			item["openid_md5_from"] = int_2_str(msgContent.openid_md5_from());
			item["openid_md5_to"] = int_2_str(msgContent.openid_md5_to());
			
			if(i == 0)
			{
				item["create_ts"] = int_2_str(msgContent.create_ts());				
			}			
			else
			{				
				if(msgContent.create_ts() - preTime >= interval)
				{					
					item["create_ts"] = int_2_str(msgContent.create_ts());
				}
				else
				{
					item["create_ts"] = "0";
				}
			}
			preTime = msgContent.create_ts();

			msg_content_list.push_front(item);
		}


		//5.Get Userinfo by openid from user server		
		::hoosho::msg::Msg stRequest;
		::hoosho::msg::MsgHead* pHead = stRequest.mutable_head();
		pHead->set_cmd(::hoosho::msg::QUERY_USER_DETAIL_INFO_REQ);
		pHead->set_seq(time(NULL));
		::hoosho::user::QueryUserDetailInfoReq* pQueryUserDetailInfoReq = stRequest.mutable_query_user_detail_info_req();
		pQueryUserDetailInfoReq->add_openid_md5_list(openid_md5_from);
		pQueryUserDetailInfoReq->add_openid_md5_list(openid_md5_to);
		
		//io
		strErrMsg = "";
		::hoosho::msg::Msg stResponse;
		::common::protoio::ProtoIOTcpClient stProtoIOTcpClient(userServerIP, userServerPort);
		iRet = stProtoIOTcpClient.io(stRequest, stResponse, strErrMsg);
		if(iRet < 0)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "Process failed, stProtoIOTcpClient.io, strErrMsg="<<strErrMsg);
			return true;	
		}

		LOG4CPLUS_DEBUG(logger, "response Msg: \n" << stResponse.Utf8DebugString());	
		//parse response
		const ::hoosho::msg::MsgHead& stHead = stResponse.head();
		if(stHead.cmd() != ::hoosho::msg::QUERY_USER_DETAIL_INFO_RES)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.cmd="<<stHead.cmd()<<", unknown, fuck!!!");
			return true;	
		}

		if(stHead.result() != ::hoosho::msg::E_OK)
		{
			DoReply(CGI_RET_CODE_SERVER_BUSY);
			LOG4CPLUS_ERROR(logger, "response.result="<<stHead.result());
			return true;	
		}

		lce::cgi::CAnyValue stHostUserInfo;
		if(stResponse.query_user_detail_info_res().user_detail_info_list_size() > 0)
		{
			const ::hoosho::commstruct::UserDetailInfo& stHostUserDetailInfo = stResponse.query_user_detail_info_res().user_detail_info_list(0);
			UserInfoPB2Any(stHostUserDetailInfo, stHostUserInfo);
		}
		

		lce::cgi::CAnyValue stGuestUserInfo;
		if(stResponse.query_user_detail_info_res().user_detail_info_list_size() > 1)
		{
			const ::hoosho::commstruct::UserDetailInfo& stGuestUserDetailInfo = stResponse.query_user_detail_info_res().user_detail_info_list(1);
			UserInfoPB2Any(stGuestUserDetailInfo, stGuestUserInfo);
		}		

		GetAnyValue()["userinfo_from"] = stHostUserInfo;
		GetAnyValue()["userinfo_to"] = stGuestUserInfo;
		GetAnyValue()["msg_content_list"] = msg_content_list;
		DoReply(CGI_RET_CODE_OK);
		return true;		
	}
		
};

int main()
{
	CgiMsgGetListBySession cgi;
	if(!cgi.Run())
	{
		return -1;
	}
	return 0;
}
